### PR TITLE
Migrate from normits

### PR DESCRIPTION
**Pull request by isaac-tfn**
# Changes

- This PR brings in some functionality from normits demand, mainly needed for caf.forecast. This is in the io.py file. Other files changed are due to linters and formatters (e.g. black).
- ...

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [ ] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [ ] Code linting, tests and other workflows pass

----
📚 Documentation preview 📚: https://caftoolkit--208.org.readthedocs.build/en/208/

### CODE REVIEW

File: src/caf/toolkit/io.py [hunk 1]
```diff
@@ -303,6 +304,84 @@ def _normalise_name(col: str) -> str:
...
+            raise FileNotFoundError(f"No such file or directory: '{path}'")
+        path = find_filename(path)
+
+    if pathlib.Path(path).suffix == ".csv":
+        return pd.read_csv(path, index_col=index_col, **kwargs)
+
+    if pathlib.Path(path).suffix in PD_COMPRESSION:
+        return pd.read_csv(path, index_col=index_col, **kwargs)
+
+    raise ValueError("Cannot determine the filetype of the given path.")
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/io.py]: I think this can be simplified. We either find a valid path in the `find_filename` that pandas can open, or we don't and the `find_filename` raises an error

```suggestion
    alt_suffix = list(PD_COMPRESSION) + [".csv"]

    # Try and find similar files if we are allowed
    if not os.path.exists(path):
        if not find_similar:
            raise FileNotFoundError(f"No such file or directory: '{path}'")
        path = find_similar_filename(path, alt_types=alt_suffix)

    if pathlib.Path(path).suffix in alt_suffix:
        return pd.read_csv(path, index_col=index_col, **kwargs)

    raise RuntimeError("A valid file was found, but apparently we can't load it??")
```

**Review by BenTaylor-TfN — COMMENTED** [review]
Added some thoughts and suggestions.

Tests are still needed for this as well.

File: src/caf/toolkit/io.py [hunk 2]
```diff
@@ -303,6 +304,84 @@ def _normalise_name(col: str) -> str:
...
+
+
+def write_df(df: pd.DataFrame, path: os.PathLike, **kwargs) -> None:
```
[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/toolkit/io.py]: I think I was just being lazy and it was easier to leave this in than refactor everything - will delete and deal with the consequences

**Review by BenTaylor-TfN — CHANGES_REQUESTED** [review]

File: src/caf/toolkit/io.py [hunk 3]
```diff
@@ -12,24 +12,34 @@
...
+    ".bz2",
+    ".zip",
+    ".xz",
+    ".zst",
+    ".tar",
+    ".tar.gz",
+    ".tar.xz",
+    ".tar.bz2",
+    ".csv.bz2",
+}
```
Comment (BenTaylor-TfN) [inline src/caf/toolkit/io.py:43]:
```suggestion
# Acceptable compression methods taken from Pandas docs: 
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html
PD_COMPRESSION = {
    ".gz",
    ".bz2",
    ".zip",
    ".xz",
    ".zst",
    ".tar",
    ".tar.gz",
    ".tar.xz",
    ".tar.bz2",
    ".csv.bz2",
}
```

**Review by BenTaylor-TfN — APPROVED** [review]
